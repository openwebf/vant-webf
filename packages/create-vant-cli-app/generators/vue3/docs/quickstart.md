# 快速上手

### 安装

```bash
# 通过 npm
npm i <%= name %>

# 通过 yarn
yarn add <%= name %>

# 通过 pnpm
pnpm add <%= name %>
```
