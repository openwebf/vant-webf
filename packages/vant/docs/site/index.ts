export { useTranslate, initDemoLocale } from './use-translate';

/** Generate the CDN URL of assets. */
export const cdnURL = (path: string) =>
  `https://fastly.jsdelivr.net/npm/@vant/assets/${path}`;
