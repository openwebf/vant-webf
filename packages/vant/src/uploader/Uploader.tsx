import {
  ref,
  reactive,
  defineComponent,
  onBeforeUnmount,
  type PropType,
  type ExtractPropTypes,
} from 'vue';

// Utils
import {
  pick,
  extend,
  toArray,
  isPromise,
  truthProp,
  Interceptor,
  getSizeStyle,
  makeArrayProp,
  makeStringProp,
  makeNumericProp,
  type Numeric,
  type ComponentInstance,
} from '../utils';
import {
  bem,
  name,
  isOversize,
  filterFiles,
  isImageFile,
  readFileContent,
} from './utils';

// Composables
import { useCustomFieldValue } from '@vant/use';
import { useExpose } from '../composables/use-expose';

// Components
import { Icon } from '../icon';
import { ImagePreview, type ImagePreviewOptions } from '../image-preview';
import UploaderPreviewItem from './UploaderPreviewItem';

// Types
import type { ImageFit } from '../image';
import type {
  UploaderExpose,
  UploaderMaxSize,
  UploaderAfterRead,
  UploaderBeforeRead,
  UploaderResultType,
  UploaderFileListItem,
} from './types';

const uploaderProps = {
  name: makeNumericProp(''),
  accept: makeStringProp('image/*'),
  capture: String,
  multiple: Boolean,
  disabled: Boolean,
  readonly: Boolean,
  lazyLoad: Boolean,
  maxCount: makeNumericProp(Infinity),
  imageFit: makeStringProp<ImageFit>('cover'),
  resultType: makeStringProp<UploaderResultType>('dataUrl'),
  uploadIcon: makeStringProp('photograph'),
  uploadText: String,
  deletable: truthProp,
  afterRead: Function as PropType<UploaderAfterRead>,
  showUpload: truthProp,
  modelValue: makeArrayProp<UploaderFileListItem>(),
  beforeRead: Function as PropType<UploaderBeforeRead>,
  beforeDelete: Function as PropType<Interceptor>,
  previewSize: [Number, String, Array] as PropType<
    Numeric | [Numeric, Numeric]
  >,
  previewImage: truthProp,
  previewOptions: Object as PropType<Partial<ImagePreviewOptions>>,
  previewFullImage: truthProp,
  maxSize: {
    type: [Number, String, Function] as PropType<UploaderMaxSize>,
    default: Infinity,
  },
};

export type UploaderProps = ExtractPropTypes<typeof uploaderProps>;

export default defineComponent({
  name,

  props: uploaderProps,

  emits: [
    'delete',
    'oversize',
    'click-upload',
    'close-preview',
    'click-preview',
    'update:modelValue',
  ],

  setup(props, { emit, slots }) {
    const inputRef = ref();
    const urls: string[] = [];

    const getDetail = (index = props.modelValue.length) => ({
      name: props.name,
      index,
    });

    const resetInput = () => {
      if (inputRef.value) {
        inputRef.value.value = '';
      }
    };

    const onAfterRead = (
      items: UploaderFileListItem | UploaderFileListItem[]
    ) => {
      resetInput();

      if (isOversize(items, props.maxSize)) {
        if (Array.isArray(items)) {
          const result = filterFiles(items, props.maxSize);
          items = result.valid;
          emit('oversize', result.invalid, getDetail());

          if (!items.length) {
            return;
          }
        } else {
          emit('oversize', items, getDetail());
          return;
        }
      }
      items = reactive(items);
      emit('update:modelValue', [...props.modelValue, ...toArray(items)]);

      if (props.afterRead) {
        props.afterRead(items, getDetail());
      }
    };

    const readFile = (files: File | File[]) => {
      const { maxCount, modelValue, resultType } = props;

      if (Array.isArray(files)) {
        const remainCount = +maxCount - modelValue.length;

        if (files.length > remainCount) {
          files = files.slice(0, remainCount);
        }

        Promise.all(
          files.map((file) => readFileContent(file, resultType))
        ).then((contents) => {
          const fileList = (files as File[]).map((file, index) => {
            const result: UploaderFileListItem = {
              file,
              status: '',
              message: '',
            };

            if (contents[index]) {
              result.content = contents[index] as string;
            }

            return result;
          });

          onAfterRead(fileList);
        });
      } else {
        readFileContent(files, resultType).then((content) => {
          const result: UploaderFileListItem = {
            file: files as File,
            status: '',
            message: '',
          };

          if (content) {
            result.content = content;
          }

          onAfterRead(result);
        });
      }
    };

    const onChange = (event: Event) => {
      const { files } = event.target as HTMLInputElement;

      if (props.disabled || !files || !files.length) {
        return;
      }

      const file =
        files.length === 1 ? files[0] : ([].slice.call(files) as File[]);

      if (props.beforeRead) {
        const response = props.beforeRead(file, getDetail());

        if (!response) {
          resetInput();
          return;
        }

        if (isPromise(response)) {
          response
            .then((data) => {
              if (data) {
                readFile(data);
              } else {
                readFile(file);
              }
            })
            .catch(resetInput);
          return;
        }
      }

      readFile(file);
    };

    let imagePreview: ComponentInstance | undefined;

    const onClosePreview = () => emit('close-preview');

    const previewImage = (item: UploaderFileListItem) => {
      if (props.previewFullImage) {
        const imageFiles = props.modelValue.filter(isImageFile);
        const images = imageFiles
          .map((item) => {
            if (item.file && !item.url && item.status !== 'failed') {
              item.url = URL.createObjectURL(item.file);
              urls.push(item.url);
            }
            return item.url;
          })
          .filter(Boolean) as string[];

        imagePreview = ImagePreview(
          extend(
            {
              images,
              startPosition: imageFiles.indexOf(item),
              onClose: onClosePreview,
            },
            props.previewOptions
          )
        );
      }
    };

    const closeImagePreview = () => {
      if (imagePreview) {
        imagePreview.close();
      }
    };

    const deleteFile = (item: UploaderFileListItem, index: number) => {
      const fileList = props.modelValue.slice(0);
      fileList.splice(index, 1);

      emit('update:modelValue', fileList);
      emit('delete', item, getDetail(index));
    };

    const renderPreviewItem = (item: UploaderFileListItem, index: number) => {
      const needPickData = [
        'imageFit',
        'deletable',
        'previewSize',
        'beforeDelete',
      ] as const;

      const previewData = extend(
        pick(props, needPickData),
        pick(item, needPickData, true)
      );

      return (
        <UploaderPreviewItem
          v-slots={pick(slots, ['preview-cover', 'preview-delete'])}
          item={item}
          index={index}
          onClick={() => emit('click-preview', item, getDetail(index))}
          onDelete={() => deleteFile(item, index)}
          onPreview={() => previewImage(item)}
          {...pick(props, ['name', 'lazyLoad'])}
          {...previewData}
        />
      );
    };

    const renderPreviewList = () => {
      if (props.previewImage) {
        return props.modelValue.map(renderPreviewItem);
      }
    };

    const onClickUpload = (event: MouseEvent) => emit('click-upload', event);

    const renderUpload = () => {
      if (props.modelValue.length >= props.maxCount || !props.showUpload) {
        return;
      }

      const Input = props.readonly ? null : (
        <input
          ref={inputRef}
          type="file"
          class={bem('input')}
          accept={props.accept}
          capture={props.capture as unknown as boolean}
          multiple={props.multiple}
          disabled={props.disabled}
          onChange={onChange}
        />
      );

      if (slots.default) {
        return (
          <div class={bem('input-wrapper')} onClick={onClickUpload}>
            {slots.default()}
            {Input}
          </div>
        );
      }

      return (
        <div
          class={bem('upload', { readonly: props.readonly })}
          style={getSizeStyle(props.previewSize)}
          onClick={onClickUpload}
        >
          <Icon name={props.uploadIcon} class={bem('upload-icon')} />
          {props.uploadText && (
            <span class={bem('upload-text')}>{props.uploadText}</span>
          )}
          {Input}
        </div>
      );
    };

    const chooseFile = () => {
      if (inputRef.value && !props.disabled) {
        inputRef.value.click();
      }
    };

    onBeforeUnmount(() => {
      urls.forEach((url) => URL.revokeObjectURL(url));
    });

    useExpose<UploaderExpose>({
      chooseFile,
      closeImagePreview,
    });
    useCustomFieldValue(() => props.modelValue);

    return () => (
      <div class={bem()}>
        <div class={bem('wrapper', { disabled: props.disabled })}>
          {renderPreviewList()}
          {renderUpload()}
        </div>
      </div>
    );
  },
});
