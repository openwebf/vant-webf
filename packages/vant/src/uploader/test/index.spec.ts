import { nextTick } from 'vue';
import { cdnURL } from '../../../docs/site';
import Uploader, { type UploaderFileListItem } from '..';
import { mount, later, triggerDrag } from '../../../test';
import type { Numeric } from '../../utils';

const mockFileDataUrl = 'data:image/test';
const mockFile = new File([new ArrayBuffer(10000)], 'test.jpg', {
  type: 'test',
});
const IMAGE = cdnURL('cat.jpeg');
const PDF = cdnURL('test.pdf');

function mockFileReader() {
  function mockReadAsText(this: FileReader) {
    if (this.onload) {
      this.onload({
        target: {
          result: mockFileDataUrl,
        },
      } as ProgressEvent<FileReader>);
    }
  }

  Object.defineProperty(window.FileReader.prototype, 'readAsText', {
    value: mockReadAsText,
  });

  Object.defineProperty(window.FileReader.prototype, 'readAsDataURL', {
    value: mockReadAsText,
  });
}

mockFileReader();

test('disabled', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    props: {
      disabled: true,
      afterRead,
    },
  });

  await later();
  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });

  input.trigger('change');
  await later();
  expect(afterRead).toHaveBeenCalledTimes(0);
});

test('result-type as text', (done) => {
  const wrapper = mount(Uploader, {
    props: {
      resultType: 'text',
      afterRead(readFile: UploaderFileListItem | UploaderFileListItem[]) {
        expect((readFile as UploaderFileListItem).content).toEqual(
          mockFileDataUrl
        );
        done();
      },
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
});

test('result-type as file', (done) => {
  const wrapper = mount(Uploader, {
    props: {
      resultType: 'file',
      afterRead: (readFile: UploaderFileListItem | UploaderFileListItem[]) => {
        expect((readFile as UploaderFileListItem).file).toBeTruthy();
        expect((readFile as UploaderFileListItem).content).toBeFalsy();
        done();
      },
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
});

test('set input name', (done) => {
  const wrapper = mount(Uploader, {
    props: {
      name: 'uploader',
      beforeRead: (
        file: File | File[],
        detail: { name: Numeric; index: number }
      ) => {
        expect(detail.name).toEqual('uploader');
        return true;
      },
      afterRead: (
        readFile: UploaderFileListItem | UploaderFileListItem[],
        detail: { name: Numeric; index: number }
      ) => {
        expect(detail.name).toEqual('uploader');
        done();
      },
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
});

test('unknown resultType', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    props: {
      afterRead,
    },
  });
  await wrapper.setProps({ resultType: 'xxxx' });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
  await later();
  expect(afterRead).toHaveBeenCalledTimes(0);
});

test('before read return false', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    props: {
      beforeRead: () => undefined,
      afterRead,
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
  await later();
  expect(afterRead).toHaveBeenCalledTimes(0);
  expect(input.element.value).toEqual('');
});

test('before read return promise and resolve', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    props: {
      beforeRead: (file: File | File[]) =>
        new Promise<File | File[] | undefined>((resolve) => {
          resolve(file);
        }),
      afterRead,
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
  await later();
  expect(afterRead).toHaveBeenCalledTimes(1);
});

test('before read return promise and resolve no value', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    props: {
      beforeRead: () =>
        new Promise<undefined>((resolve) => {
          resolve(undefined);
        }),
      afterRead,
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
  await later();
  expect(afterRead).toHaveBeenCalledTimes(1);
  expect(input.element.value).toEqual('');
});

test('before read return promise and reject', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    props: {
      beforeRead: () =>
        new Promise<undefined>((resolve, reject) => {
          reject();
        }),
      afterRead,
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
  await later();
  expect(afterRead).toHaveBeenCalledTimes(0);
  expect(input.element.value).toEqual('');
});

test('should trigger oversize event when file size is overlimit', async () => {
  const wrapper = mount(Uploader, {
    props: {
      maxSize: 1,
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');

  let fileList = [mockFile];
  Object.defineProperty(input.element, 'files', {
    get: () => jest.fn().mockReturnValue(fileList)(),
  });

  await input.trigger('change');

  fileList = [mockFile, mockFile];
  await input.trigger('change');

  await later();
  expect(wrapper.emitted<[File]>('oversize')![0]).toBeTruthy();
  expect(wrapper.emitted<[File]>('oversize')![1]).toBeTruthy();

  await wrapper.setProps({ maxSize: 100000 });

  fileList = [mockFile];
  await input.trigger('change');

  await later();
  expect(wrapper.emitted<[File]>('oversize')![2]).toBeFalsy();
});

test('should allow to custom max-size for different type of files', async () => {
  const wrapper = mount(Uploader, {
    props: {
      maxSize(file: File) {
        if (file.type === 'test') {
          return file.size > 500;
        }
        return false;
      },
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  const fileList = [mockFile];

  Object.defineProperty(input.element, 'files', {
    get: () => jest.fn().mockReturnValue(fileList)(),
  });

  await input.trigger('change');

  await later();
  expect(wrapper.emitted<[File]>('oversize')![0]).toBeTruthy();
});

test('render upload-text', () => {
  const uploadText = 'Text';
  const wrapper = mount(Uploader, {
    props: {
      uploadText,
    },
  });

  const text = wrapper.find<HTMLSpanElement>('.van-uploader__upload-text');
  expect(text?.text()).toEqual(uploadText);
});

test('render preview image', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [
        { url: cdnURL('cat.jpeg') },
        { url: cdnURL('test.pdf') },
        { file: mockFile },
      ],
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('image-fit prop', () => {
  const wrapper = mount(Uploader, {
    props: {
      imageFit: 'contain',
      modelValue: [{ url: cdnURL('cat.jpeg') }],
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('upload-icon prop', () => {
  const wrapper = mount(Uploader, {
    props: {
      uploadIcon: 'add',
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('disable preview image', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [],
      previewImage: false,
    },
  });

  await wrapper.setProps({ modelValue: [{ file: mockFile }] });
  expect(wrapper.html()).toMatchSnapshot();

  await wrapper.setProps({ previewImage: true });
  expect(wrapper.html()).toMatchSnapshot();
});

test('max-count prop', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [],
      maxCount: 1,
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile, mockFile]),
  });
  input.trigger('change');
  await later();
  expect(
    wrapper.emitted<[File | File[]]>('update:modelValue')![0][0]
  ).toHaveLength(1);
});

test('should allow to custom size by preview-size prop', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ file: mockFile }],
      previewSize: 30,
    },
  });

  const image = wrapper.find('.van-uploader__file');
  expect(image.style.width).toEqual('30px');
  expect(image.style.height).toEqual('30px');

  const upload = wrapper.find('.van-uploader__upload');
  expect(upload.style.width).toEqual('30px');
  expect(upload.style.height).toEqual('30px');
});

test('should allow to set width and height separately by preview-size prop', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ file: mockFile }],
      previewSize: [20, 10],
    },
  });

  const image = wrapper.find('.van-uploader__file');
  expect(image.style.width).toEqual('20px');
  expect(image.style.height).toEqual('10px');

  const upload = wrapper.find('.van-uploader__upload');
  expect(upload.style.width).toEqual('20px');
  expect(upload.style.height).toEqual('10px');
});

test('deletable prop', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
    },
  });

  expect(wrapper.find('.van-uploader__preview-delete').exists()).toBeTruthy();

  await wrapper.setProps({ deletable: false });
  expect(wrapper.find('.van-uploader__preview-delete').exists()).toBeFalsy();
});

test('delete preview image', () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
      previewSize: 30,
    },
  });

  wrapper.find('.van-uploader__preview-delete').trigger('click');

  expect(wrapper.emitted<[File]>('update:modelValue')![0][0]).toHaveLength(0);
  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.emitted<[File]>('delete')![0]).toBeTruthy();
});

test('before-delete prop return false', () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
      beforeDelete: () => false,
    },
  });

  wrapper.find('.van-uploader__preview-delete').trigger('click');
  expect(wrapper.emitted('delete')).toBeFalsy();
});

test('before-delete prop return true', () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
      beforeDelete: () => true,
    },
  });

  wrapper.find('.van-uploader__preview-delete').trigger('click');
  expect(wrapper.emitted('delete')).toBeTruthy();
});

test('before-delete prop resolved', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
      beforeDelete: () =>
        new Promise<boolean>((resolve) => {
          resolve(true);
        }),
    },
  });

  wrapper.find('.van-uploader__preview-delete').trigger('click');
  await later();
  expect(wrapper.emitted('delete')).toBeTruthy();
});

test('before-delete prop rejected', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
      beforeDelete: () =>
        new Promise<boolean>((resolve, reject) => {
          reject();
        }),
    },
  });

  wrapper.find('.van-uploader__preview-delete').trigger('click');
  await later();
  expect(wrapper.emitted('delete')).toBeFalsy();
});

test('click to preview image', async () => {
  const wrapper = mount(Uploader, {
    props: {
      previewFullImage: false,
      modelValue: [{ url: IMAGE }, { url: PDF }],
    },
  });

  wrapper.find('.van-image').trigger('click');
  await later();
  expect(document.querySelector('.van-image-preview')).toBeFalsy();

  await wrapper.setProps({ previewFullImage: true });
  wrapper.find('.van-image').trigger('click');
  await later();
  expect(document.querySelector('.van-image-preview')).toBeTruthy();

  const images = document.querySelectorAll<HTMLImageElement>(
    '.van-image-preview .van-image-preview__image'
  );
  expect(images).toHaveLength(1);
});

test('preview-options prop', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
      previewOptions: {
        images: [],
        closeable: true,
      },
    },
  });

  wrapper.find('.van-image').trigger('click');
  await later();

  const closeIcon = document.querySelectorAll('.van-image-preview__close-icon');
  expect(closeIcon).toHaveLength(1);
});

test('closeImagePreview method', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
    },
  });

  await nextTick();
  await (wrapper.vm as Record<string, any>).closeImagePreview();
  expect(wrapper.emitted('close-preview')).toBeFalsy();

  wrapper.find('.van-image').trigger('click');
  await (wrapper.vm as Record<string, any>).closeImagePreview();
  expect(wrapper.emitted('close-preview')).toBeTruthy();
});

test('click-preview event', () => {
  const wrapper = mount(Uploader, {
    props: {
      previewFullImage: false,
      modelValue: [{ url: IMAGE }, { url: PDF }],
    },
  });

  wrapper.find('.van-image').trigger('click');
  expect(wrapper.emitted<[File]>('click-preview')![0][0]).toEqual({
    url: IMAGE,
  });
  expect(
    wrapper.emitted<[File, { name: string; index: number }]>(
      'click-preview'
    )![0][1]
  ).toEqual({
    name: '',
    index: 0,
  });
});

test('close-preview event', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
    },
  });

  await later();
  await wrapper.find('.van-image').trigger('click');

  const preview = document.querySelector<HTMLDivElement>('.van-image-preview');
  const swipe = preview?.querySelector<HTMLDivElement>(
    '.van-swipe-item'
  ) as HTMLDivElement;
  triggerDrag(swipe, 0, 0);

  await later(300);
  expect(wrapper.emitted('close-preview')).toBeTruthy();
});

test('show-upload prop', async () => {
  const wrapper = mount(Uploader);

  expect(wrapper.find('.van-uploader__upload').exists()).toBeTruthy();
  await wrapper.setProps({ showUpload: false });
  expect(wrapper.find('.van-uploader__upload').exists()).toBeFalsy();
});

test('file message should be reactive', (done) => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [],
      afterRead(file: UploaderFileListItem | UploaderFileListItem[]) {
        if (Array.isArray(file)) return;
        file.status = 'uploading';
        file.message = '1';
        setTimeout(() => {
          file.message = '2';
          expect(wrapper.html()).toMatchSnapshot();
          done();
        });
      },
    },
  });

  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile]),
  });
  input.trigger('change');
});

test('multiFile upload filter max-size file', async () => {
  const wrapper = mount(Uploader, {
    props: {
      maxSize: 1000,
    },
  });

  const smallFile = new File([new ArrayBuffer(100)], 'small.jpg');
  const input = wrapper.find<HTMLInputElement>('.van-uploader__input');
  Object.defineProperty(input.element, 'files', {
    get: jest.fn().mockReturnValue([mockFile, smallFile]),
  });
  input.trigger('change');
  await later();

  expect(wrapper.emitted<[File]>('oversize')![0]).toBeTruthy();
});

test('should render preview-cover slot correctly', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
    },
    slots: {
      'preview-cover': 'Custom Preview Cover',
    },
  });

  expect(wrapper.find('.van-uploader__preview-cover').html()).toMatchSnapshot();
});

test('should render preview-delete slot correctly', async () => {
  const wrapper = mount(Uploader, {
    props: {
      modelValue: [{ url: IMAGE }],
    },
    slots: {
      'preview-delete': 'Custom Preview Delete',
    },
  });

  expect(
    wrapper.find('.van-uploader__preview-delete').html()
  ).toMatchSnapshot();
});

test('should not render upload input when using readonly prop', async () => {
  const wrapper = mount(Uploader, {
    props: {
      readonly: true,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should emit click-upload event when upload area is clicked', async () => {
  const wrapper = mount(Uploader);
  wrapper.find('.van-uploader__upload').trigger('click');
  expect(wrapper.emitted('click-upload')).toBeTruthy();
});
