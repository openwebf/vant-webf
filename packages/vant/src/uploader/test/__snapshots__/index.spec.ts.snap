// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete preview image 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__preview">
      <div class="van-image van-uploader__preview-image"
           style="width: 30px; height: 30px;"
      >
        <img src="https://fastly.jsdelivr.net/npm/@vant/assets/cat.jpeg"
             class="van-image__img"
             style="object-fit: cover;"
        >
        <div class="van-image__loading">
          <i class="van-badge__wrapper van-icon van-icon-photo van-image__loading-icon">
          </i>
        </div>
      </div>
      <div role="button"
           class="van-uploader__preview-delete van-uploader__preview-delete--shadow"
           tabindex="0"
           aria-label="Delete"
      >
        <i class="van-badge__wrapper van-icon van-icon-cross van-uploader__preview-delete-icon">
        </i>
      </div>
    </div>
    <div class="van-uploader__upload"
         style="width: 30px; height: 30px;"
    >
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;

exports[`disable preview image 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__upload">
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;

exports[`disable preview image 2`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__preview">
      <div class="van-uploader__file">
        <i class="van-badge__wrapper van-icon van-icon-description van-uploader__file-icon">
        </i>
        <div class="van-uploader__file-name van-ellipsis">
          test.jpg
        </div>
      </div>
      <div role="button"
           class="van-uploader__preview-delete van-uploader__preview-delete--shadow"
           tabindex="0"
           aria-label="Delete"
      >
        <i class="van-badge__wrapper van-icon van-icon-cross van-uploader__preview-delete-icon">
        </i>
      </div>
    </div>
    <div class="van-uploader__upload">
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;

exports[`file message should be reactive 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__upload">
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;

exports[`image-fit prop 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__preview">
      <div class="van-image van-uploader__preview-image">
        <img src="https://fastly.jsdelivr.net/npm/@vant/assets/cat.jpeg"
             class="van-image__img"
             style="object-fit: contain;"
        >
        <div class="van-image__loading">
          <i class="van-badge__wrapper van-icon van-icon-photo van-image__loading-icon">
          </i>
        </div>
      </div>
      <div role="button"
           class="van-uploader__preview-delete van-uploader__preview-delete--shadow"
           tabindex="0"
           aria-label="Delete"
      >
        <i class="van-badge__wrapper van-icon van-icon-cross van-uploader__preview-delete-icon">
        </i>
      </div>
    </div>
    <div class="van-uploader__upload">
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;

exports[`render preview image 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__preview">
      <div class="van-image van-uploader__preview-image">
        <img src="https://fastly.jsdelivr.net/npm/@vant/assets/cat.jpeg"
             class="van-image__img"
             style="object-fit: cover;"
        >
        <div class="van-image__loading">
          <i class="van-badge__wrapper van-icon van-icon-photo van-image__loading-icon">
          </i>
        </div>
      </div>
      <div role="button"
           class="van-uploader__preview-delete van-uploader__preview-delete--shadow"
           tabindex="0"
           aria-label="Delete"
      >
        <i class="van-badge__wrapper van-icon van-icon-cross van-uploader__preview-delete-icon">
        </i>
      </div>
    </div>
    <div class="van-uploader__preview">
      <div class="van-uploader__file">
        <i class="van-badge__wrapper van-icon van-icon-description van-uploader__file-icon">
        </i>
        <div class="van-uploader__file-name van-ellipsis">
          https://fastly.jsdelivr.net/npm/@vant/assets/test.pdf
        </div>
      </div>
      <div role="button"
           class="van-uploader__preview-delete van-uploader__preview-delete--shadow"
           tabindex="0"
           aria-label="Delete"
      >
        <i class="van-badge__wrapper van-icon van-icon-cross van-uploader__preview-delete-icon">
        </i>
      </div>
    </div>
    <div class="van-uploader__preview">
      <div class="van-uploader__file">
        <i class="van-badge__wrapper van-icon van-icon-description van-uploader__file-icon">
        </i>
        <div class="van-uploader__file-name van-ellipsis">
          test.jpg
        </div>
      </div>
      <div role="button"
           class="van-uploader__preview-delete van-uploader__preview-delete--shadow"
           tabindex="0"
           aria-label="Delete"
      >
        <i class="van-badge__wrapper van-icon van-icon-cross van-uploader__preview-delete-icon">
        </i>
      </div>
    </div>
    <div class="van-uploader__upload">
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;

exports[`should not render upload input when using readonly prop 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__upload van-uploader__upload--readonly">
      <i class="van-badge__wrapper van-icon van-icon-photograph van-uploader__upload-icon">
      </i>
    </div>
  </div>
</div>
`;

exports[`should render preview-cover slot correctly 1`] = `
<div class="van-uploader__preview-cover">
  Custom Preview Cover
</div>
`;

exports[`should render preview-delete slot correctly 1`] = `
<div role="button"
     class="van-uploader__preview-delete"
     tabindex="0"
     aria-label="Delete"
>
  Custom Preview Delete
</div>
`;

exports[`upload-icon prop 1`] = `
<div class="van-uploader">
  <div class="van-uploader__wrapper">
    <div class="van-uploader__upload">
      <i class="van-badge__wrapper van-icon van-icon-add van-uploader__upload-icon">
      </i>
      <input type="file"
             class="van-uploader__input"
             accept="image/*"
      >
    </div>
  </div>
</div>
`;
